File: AppLovin/SASAppLovinRewardedVideoAdapter.m
//
//  SASAppLovinRewardedVideoAdapter.m
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 24/09/2018.
//  Copyright © 2018 Smart AdServer. All rights reserved.
//

#import "SASAppLovinRewardedVideoAdapter.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SASAppLovinRewardedVideoAdapter

#pragma mark - Mediation adapter protocol implementation

- (instancetype)initWithDelegate:(id<SASMediationRewardedVideoAdapterDelegate>)delegate {
    if (self = [super init]) {
        self.delegate = delegate;
    }
    return self;
}

- (void)requestRewardedVideoWithServerParameterString:(NSString *)serverParameterString clientParameters:(NSDictionary *)clientParameters {
    self.isReady = NO;
    
    // Initializing the SDK is done in the base class
    [SASAppLovinBaseAdapter initializeAppLovin];
    
    // Configuring GDPR is done in the base class
    [self configureGDPRWithClientParameters:clientParameters];
    
    // Instantiating & loading a rewarded video ad
    self.rewardedVideo = [[ALIncentivizedInterstitialAd alloc] initWithSdk:[ALSdk shared]];
    self.rewardedVideo.adDisplayDelegate = self;
    [self.rewardedVideo preloadAndNotify:self];
}

- (void)showRewardedVideoFromViewController:(UIViewController *)viewController {
    [self.rewardedVideo showAndNotify:self];
}

- (BOOL)isRewardedVideoReady {
    return self.isReady;
}

#pragma mark - AppLovin rewarded video delegate

- (void)adService:(ALAdService *)adService didLoadAd:(ALAd *)ad {
    self.isReady = YES;
    [self.delegate mediationRewardedVideoAdapterDidLoad:self];
}

- (void)adService:(ALAdService *)adService didFailToLoadAdWithError:(int)code {
    NSError *error = [NSError errorWithDomain:SASAppLovinAdapterErrorDomain code:code userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"AppLovin RewardedVideo Error: %ld", (long)code] }];
    [self.delegate mediationRewardedVideoAdapter:self didFailToLoadWithError:error noFill:(code == kALErrorCodeNoFill)];
}

- (void)ad:(ALAd *)ad wasDisplayedIn:(UIView *)view {
    [self.delegate mediationRewardedVideoAdapterDidShow:self];
}

- (void)ad:(ALAd *)ad wasHiddenIn:(UIView *)view {
    self.isReady = NO;
    [self.delegate mediationRewardedVideoAdapterDidClose:self];
}

- (void)ad:(ALAd *)ad wasClickedIn:(UIView *)view {
    [self.delegate mediationRewardedVideoAdapterDidReceiveAdClickedEvent:self];
}

- (void)rewardValidationRequestForAd:(ALAd *)ad didSucceedWithResponse:(NSDictionary *)response {
    // A reward is provided by AppLovin, we can use it instead of relying on a Smart reward.
    NSString *currency = [response objectForKey: @"currency"];
    NSNumber *amount = [NSNumber numberWithFloat:[[response objectForKey: @"amount"] floatValue]];
    
    SASReward *reward = [[SASReward alloc] initWithAmount:amount currency:currency];
    [self.delegate mediationRewardedVideoAdapter:self didCollectReward:reward];
}

- (void)rewardValidationRequestForAd:(ALAd *)ad didExceedQuotaWithResponse:(NSDictionary *)response {
    // If the reward fails for any reason, the adapter should fail to show with a relevant error
    self.isReady = NO;
    NSError *error = [NSError errorWithDomain:SASAppLovinAdapterErrorDomain code:SASAppLovinAdapterErrorCodeUnableToGetReward userInfo:@{ NSLocalizedDescriptionKey: SASAppLovinAdapterErrorMessageUnableToGetReward }];
    [self.delegate mediationRewardedVideoAdapter:self didFailToShowWithError:error];
}

- (void)rewardValidationRequestForAd:(ALAd *)ad wasRejectedWithResponse:(NSDictionary *)response {
    // If the reward fails for any reason, the adapter should fail to show with a relevant error
    self.isReady = NO;
    NSError *error = [NSError errorWithDomain:SASAppLovinAdapterErrorDomain code:SASAppLovinAdapterErrorCodeUnableToGetReward userInfo:@{ NSLocalizedDescriptionKey: SASAppLovinAdapterErrorMessageUnableToGetReward }];
    [self.delegate mediationRewardedVideoAdapter:self didFailToShowWithError:error];
}

- (void)rewardValidationRequestForAd:(ALAd *)ad didFailWithError:(NSInteger)responseCode {
    // If the reward fails for any reason, the adapter should fail to show with a relevant error
    self.isReady = NO;
    NSError *error = [NSError errorWithDomain:SASAppLovinAdapterErrorDomain code:SASAppLovinAdapterErrorCodeUnableToGetReward userInfo:@{ NSLocalizedDescriptionKey: SASAppLovinAdapterErrorMessageUnableToGetReward }];
    [self.delegate mediationRewardedVideoAdapter:self didFailToShowWithError:error];
}

- (void)userDeclinedToViewAd:(ALAd *)ad {
    // If the reward fails for any reason, the adapter should fail to show with a relevant error
    self.isReady = NO;
    NSError *error = [NSError errorWithDomain:SASAppLovinAdapterErrorDomain code:SASAppLovinAdapterErrorCodeUserDeclinedToViewAd userInfo:@{ NSLocalizedDescriptionKey: SASAppLovinAdapterErrorMessageUserDeclinedToViewAd }];
    [self.delegate mediationRewardedVideoAdapter:self didFailToShowWithError:error];
}

@end

NS_ASSUME_NONNULL_END

File: AppLovin/SASAppLovinRewardedVideoAdapter.swift
//
//  SASAppLovinRewardedVideoAdapter.swift
//  AdViewer2-Mediation
//
//  Created by Loïc GIRON DIT METAZ on 11/02/2022.
//  Copyright © 2022 Smart AdServer. All rights reserved.
//

import Foundation
import AppLovinSDK
import SASDisplayKit

/**
 AppLovin rewarded video adapter class for Smart SDK mediation.
 
 Mediation adapter classes can be used to display an ad using a third party SDK directly from
 an insertion handled by Smart.
 
 To use an adapter class, you simply have to add them to your Xcode project and they will
 be automatically instantiated by the Smart SDK if needed.
 */
@objc(SASAppLovinRewardedVideoAdapter)
class SASAppLovinRewardedVideoAdapter: SASAppLovinBaseAdapter, SASMediationRewardedVideoAdapter, MARewardedAdDelegate {
   
    
    /// A delegate that this adapter must call to provide information about the ad loading status or events to the Smart SDK.
    private weak var delegate: SASMediationRewardedVideoAdapterDelegate? = nil

    /// The currently loaded AppLovin rewarded video if any.
    private var rewardedVideo: MARewardedAd? = nil

    /// YES if an ad is ready to be displayed, NO otherwise.
    private var isReady = false
    
    // MARK: - Mediation adapter protocol implementation
    
    required init(delegate: SASMediationRewardedVideoAdapterDelegate) {
        self.delegate = delegate
    }
    
    func requestRewardedVideo(withServerParameterString serverParameterString: String, clientParameters: [AnyHashable : Any]) {
        isReady = false
        
        // Parameter retrieval and validation
        do {
            let (adUnitID, _) = try configureAppLovinAds(serverParameterString: serverParameterString)
            // Initializing the SDK is done in the base class
            SASAppLovinBaseAdapter.initializeApplovinSdk { [self] in
                // Configuring GDPR is done in the base class
                configureGDPRWithClientParameters(clientParameters: clientParameters)
                
                // Instantiating & loading a rewarded video ad
                let rewardedVideo = MARewardedAd.shared(withAdUnitIdentifier: adUnitID)
                rewardedVideo.delegate = self
                
                // Load the ad
                rewardedVideo.load()
                
                self.rewardedVideo = rewardedVideo
            }
        } catch {
            delegate?.mediationRewardedVideoAdapter(self, didFailToLoadWithError: error, noFill: false)
        }
    }
    
    func showRewardedVideo(from viewController: UIViewController) {
        rewardedVideo?.show(forPlacement: nil, customData: nil, viewController: viewController)
    }
    
    func isRewardedVideoReady() -> Bool {
        return rewardedVideo?.isReady ?? false
    }
    
    // MARK: - AppLovin delegates
    
    func didStartRewardedVideo(for ad: MAAd) {
        // nothing to do
    }
    
    func didCompleteRewardedVideo(for ad: MAAd) {
        // nothing to do
    }
    
    func didRewardUser(for ad: MAAd, with reward: MAReward) {
        // A reward is provided by AppLovin, we can use it instead of relying on a Smart reward.
        let currency = reward.label
        let amount = NSNumber(floatLiteral: Double(reward.amount))
        
        let reward = SASReward(amount: amount, currency: currency)
        delegate?.mediationRewardedVideoAdapter(self, didCollect: reward)
    }
    
    func didLoad(_ ad: MAAd) {
        delegate?.mediationRewardedVideoAdapterDidLoad(self)
    }
    
    func didFailToLoadAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {
        let error = NSError(domain: SASAppLovinBaseAdapter.ErrorConstants.errorDomain, code: error.code.rawValue, userInfo: [
            NSLocalizedDescriptionKey: "AppLovin Rewarded Error: \(error.code.rawValue)"
        ])
        delegate?.mediationRewardedVideoAdapter(self, didFailToLoadWithError: error, noFill: MAErrorCode.noFill.rawValue == error.code)
    }
    
    func didDisplay(_ ad: MAAd) {
        delegate?.mediationRewardedVideoAdapterDidShow(self)
    }
    
    func didHide(_ ad: MAAd) {
        delegate?.mediationRewardedVideoAdapterDidClose(self)
    }
    
    func didClick(_ ad: MAAd) {
        delegate?.mediationRewardedVideoAdapterDidReceiveAdClickedEvent(self)
    }
    
    func didFail(toDisplay ad: MAAd, withError error: MAError) {
        let error = NSError(domain: SASAppLovinBaseAdapter.ErrorConstants.errorDomain, code: error.code.rawValue, userInfo: [
            NSLocalizedDescriptionKey: "AppLovin Rewarded Error: \(error.code.rawValue)"
        ])
        delegate?.mediationRewardedVideoAdapter(self, didFailToShowWithError: error)
    }
    

}
